gateway-js/src/datasources/types.ts

```diff
@@ -1,7 +1,48 @@
 import { GraphQLResponse, GraphQLRequestContext } from 'apollo-server-types';
 
-export interface GraphQLDataSource<TContext extends Record<string, any> = Record<string, any>> {
+export interface GraphQLDataSource<
+  TContext extends Record<string, any> = Record<string, any>,
+> {
   process(
-    request: Pick<GraphQLRequestContext<TContext>, 'request' | 'context'>,
+    options: GraphQLDataSourceProcessOptions<TContext>,
   ): Promise<GraphQLResponse>;
 }
+
+export type GraphQLDataSourceProcessOptions<
+  TContext extends Record<string, any> = Record<string, any>,
+> = {
+  /**
+   * The request to send to the subgraph.
+   */
+  request: GraphQLRequestContext<TContext>['request'];
+} & (
+  | {
+      kind: 'incoming operation';
+      /**
+       * The GraphQLRequestContext for the operation received by the gateway, or
+       * one of the strings if this operation is generated by the gateway without an
+       * incoming request.
+       */
+      incomingRequestContext: GraphQLRequestContext<TContext>;
+      /**
+       * Equivalent to incomingRequestContext.context (provided here for
+       * backwards compatibility): the object created by the Apollo Server
+       * `context` function.
+       */
+      context: GraphQLRequestContext<TContext>['context'];
+    }
+  | {
+      kind: 'health check';
```

Comment:
This is maybe just a bit off of convention. Tempting to make this an enum or enum-esque a la graphql's `Kind` https://github.com/graphql/graphql-js/blob/main/src/language/kinds.ts

That being said, it's not like TS struggles at all with these string types - it'll auto-complete and type-check these `kind` properties just fine as-is.

Reply:
OK, enum it is.